### PR TITLE
[Merged by Bors] - feat: explicit formula for the complex square root

### PR DESCRIPTION
**Pull request by urkud**
Also add formula for `cos (x / 2)`, `sin (x / 2)` and golf some proofs.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/SpecialFunctions/Pow/Real.lean [hunk 1]
```diff
@@ -255,6 +255,22 @@ theorem ofReal_cpow_of_nonpos {x : ℝ} (hx : x ≤ 0) (y : ℂ) :
     arg_ofReal_of_nonneg (neg_nonneg.2 hx), ofReal_zero, zero_mul, add_zero]
 #align complex.of_real_cpow_of_nonpos Complex.ofReal_cpow_of_nonpos
 
+lemma cpow_ofReal (x : ℂ) (y : ℝ) :
+    x ^ y = ↑((abs x) ^ y) * (Real.cos (arg x * y) + Real.sin (arg x * y) * I) := by
```
Comment (eric-wieser) [inline Mathlib/Analysis/SpecialFunctions/Pow/Real.lean:259]: Can you add explicit casts here? This looks very fragile WRT the elaboration of `^`

Comment (eric-wieser) [reply Mathlib/Analysis/SpecialFunctions/Pow/Real.lean:259]:
```suggestion
    x ^ (y : ℝ) = ↑(abs x ^ y) * (Real.cos (arg x * y) + Real.sin (arg x * y) * I) := by
```

File: Mathlib/Analysis/SpecialFunctions/Pow/Real.lean [hunk 2]
```diff
@@ -255,6 +255,22 @@ theorem ofReal_cpow_of_nonpos {x : ℝ} (hx : x ≤ 0) (y : ℂ) :
...
+  rcases eq_or_ne x 0 with rfl | hx
+  · simp [ofReal_cpow le_rfl]
+  · rw [cpow_def_of_ne_zero hx, exp_eq_exp_re_mul_sin_add_cos, mul_comm (log x)]
+    norm_cast
+    rw [ofReal_mul_re, ofReal_mul_im, log_re, log_im, mul_comm y, mul_comm y, Real.exp_mul,
+      Real.exp_log]
+    rwa [abs.pos_iff]
+
+lemma cpow_ofReal_re (x : ℂ) (y : ℝ) : (x ^ y).re = (abs x) ^ y * Real.cos (arg x * y) := by
```
Comment (eric-wieser) [inline Mathlib/Analysis/SpecialFunctions/Pow/Real.lean:268]:
```suggestion
lemma cpow_ofReal_re (x : ℂ) (y : ℝ) : (x ^ (y : ℝ)).re = (abs x) ^ y * Real.cos (arg x * y) := by
```

File: Mathlib/Analysis/SpecialFunctions/Pow/Real.lean [hunk 3]
```diff
@@ -764,6 +780,20 @@ theorem exists_rat_pow_btwn {α : Type*} [LinearOrderedField α] [Archimedean α
...
 end Real
 
+namespace Complex
+
+lemma cpow_inv_two_re (x : ℂ) : (x ^ (2⁻¹ : ℂ)).re = sqrt ((abs x + x.re) / 2) := by
+  rw [← ofReal_ofNat, ← ofReal_inv, cpow_ofReal_re, ← div_eq_mul_inv, ← one_div,
+    ← Real.sqrt_eq_rpow, cos_half, ← sqrt_mul, ← mul_div_assoc, mul_add, mul_one, abs_mul_cos_arg]
+  exacts [abs.nonneg _, (neg_pi_lt_arg _).le, arg_le_pi _]
+
+lemma abs_cpow_inv_two_im (x : ℂ) : |(x ^ (2⁻¹ : ℂ)).im| = sqrt ((abs x - x.re) / 2) := by
```
Comment (eric-wieser) [inline Mathlib/Analysis/SpecialFunctions/Pow/Real.lean:790]: is this true as follows?
```suggestion
lemma abs_cpow_inv_two_im (x : ℂ) : (x ^ (2⁻¹ : ℂ)).im = x.im.sign * sqrt ((abs x - x.re) / 2) := by
```

Comment (urkud) [reply Mathlib/Analysis/SpecialFunctions/Pow/Real.lean:790]: No, this isn't true for real negative `x`.

Comment (eric-wieser) [reply Mathlib/Analysis/SpecialFunctions/Pow/Real.lean:790]: How about separate lemmas for `0 ≤ x.im` and `x.im < 0` (in addition to the one you already have)

File: Mathlib/Analysis/SpecialFunctions/Trigonometric/Basic.lean [hunk 4]
```diff
@@ -503,6 +503,23 @@ theorem cos_eq_sqrt_one_sub_sin_sq {x : ℝ} (hl : -(π / 2) ≤ x) (hu : x ≤
   rw [← abs_cos_eq_sqrt_one_sub_sin_sq, abs_of_nonneg (cos_nonneg_of_mem_Icc ⟨hl, hu⟩)]
 #align real.cos_eq_sqrt_one_sub_sin_sq Real.cos_eq_sqrt_one_sub_sin_sq
 
+lemma cos_half {x : ℝ} (hl : -π ≤ x) (hr : x ≤ π) : cos (x / 2) = sqrt ((1 + cos x) / 2) := by
```
Comment (eric-wieser) [inline Mathlib/Analysis/SpecialFunctions/Trigonometric/Basic.lean:506]:
```suggestion
lemma cos_div_two {x : ℝ} (hl : -π ≤ x) (hr : x ≤ π) : cos (x / 2) = sqrt ((1 + cos x) / 2) := by
```
seems better aligned with the naming scheme (and below)

Comment (eric-wieser) [reply Mathlib/Analysis/SpecialFunctions/Trigonometric/Basic.lean:506]: (unless there's already precedent that I've missed)

**Review by eric-wieser — APPROVED** [review]
bors d+

The `half` naming worries me slightly, but otherwise this looks fine. The extended version of `abs_cpow_inv_two_im` is only in scope if you want it to be: though maybe a TODO comment would be reasonable if you decide it's not.

File: Mathlib/Analysis/SpecialFunctions/Trigonometric/Basic.lean (hunk 4 above)
Comment (urkud) [reply Mathlib/Analysis/SpecialFunctions/Trigonometric/Basic.lean:506]: Precedents: https://loogle.lean-lang.org/?q=%22half%22 https://loogle.lean-lang.org/?q=%22halves%22

**Comment by urkud** [comment]
bors merge

**Pull request closed without merging**